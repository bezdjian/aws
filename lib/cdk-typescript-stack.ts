import * as lambda from '@aws-cdk/aws-lambda';
import * as api from '@aws-cdk/aws-apigateway';
import * as cdk from '@aws-cdk/core';
import * as path from 'path';

export class CdkTypescriptStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // The code that defines your stack goes here
    const fn = new lambda.Function(this, "HelloFn", {
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, 'lambdafunction')),
      functionName: 'CdkHelloFn'
    });

    const gateway = new api.LambdaRestApi(this, "HelloFnApi", {
      handler: fn,
      restApiName: 'HelloFnApi',
      description: 'Api gateway created by CDK for HelloFn Lambda',
      proxy: false
    });

    gateway.root.addMethod('GET')
  }
}
