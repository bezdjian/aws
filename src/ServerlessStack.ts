import * as lambda from '@aws-cdk/aws-lambda';
import * as api from '@aws-cdk/aws-apigateway';
import * as cdk from '@aws-cdk/core';
import { AttributeType, Table } from '@aws-cdk/aws-dynamodb';
import { CfnOutput } from '@aws-cdk/core';

export class ServerlessStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create DynamoDB
    const tableName = 'cdkTable'
    let dynamoTable = new Table(this, "CdkDynamoDBTable", {
      tableName,
      partitionKey: {name: 'id', type: AttributeType.STRING},
      removalPolicy: cdk.RemovalPolicy.DESTROY
    });

    // Create Lambda function with an existing handler
    const fn = new lambda.Function(this, "HelloFn", {
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset('./resource'),
      functionName: 'CdkHelloFn',
      environment: {
        DB_TABLE: dynamoTable.tableName
      }
    });
    
    // Grant read access to lambda function
    dynamoTable.grantReadData(fn)

    // Create API Gateway with GET method triggering Lambda function
    const gateway = new api.LambdaRestApi(this, "HelloFnApi", {
      handler: fn,
      restApiName: 'HelloFnApi',
      description: 'Api gateway created by CDK for HelloFn Lambda',
      proxy: false
    });

    gateway.root.resourceForPath('scan').addMethod('GET', new api.LambdaIntegration(fn));

    // Expose api url to output
    new CfnOutput(this, 'ScanApiGatewayUrl', {
      value: gateway.url,
      exportName: 'ScanApiGatewayUrl'
    })
  }
}
