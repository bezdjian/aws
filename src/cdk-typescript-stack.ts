import * as lambda from '@aws-cdk/aws-lambda';
import * as api from '@aws-cdk/aws-apigateway';
import * as cdk from '@aws-cdk/core';
import * as path from 'path';
import { AttributeType, Table } from '@aws-cdk/aws-dynamodb';
import { CfnOutput } from '@aws-cdk/core';

export class CdkTypescriptStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create DynamoDB
    const tableName = 'cdkTable'
    let dynamoTable = new Table(this, "CdkDynamoDBTable", {
      tableName,
      partitionKey: {name: 'id', type: AttributeType.STRING},
      removalPolicy: cdk.RemovalPolicy.DESTROY
    });

    // Create Lambda function with an existing handler
    const fn = new lambda.Function(this, "HelloFn", {
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, 'lambdafunction')),
      functionName: 'CdkHelloFn',
      environment: {
        DB_TABLE: dynamoTable.tableName
      }
    });
    //OR. Add DynamoDB table name to Lambda environment variable
    //fn.addEnvironment('DBTABLE', dynamoTable.tableName);

    // Grant read access to lambda function
    dynamoTable.grantReadData(fn)

    // Create API Gateway with GET method triggering Lambda function
    const gateway = new api.LambdaRestApi(this, "HelloFnApi", {
      handler: fn,
      restApiName: 'HelloFnApi',
      description: 'Api gateway created by CDK for HelloFn Lambda',
      proxy: false
    });

    gateway.root.addMethod('GET')
    gateway.root.addResource('scan')

    // Expose api url to output
    new CfnOutput(this, 'ScanApiGatewayUrl', {
      value: gateway.url,
      exportName: 'ScanApiGatewayUrl'
    })
  }
}
