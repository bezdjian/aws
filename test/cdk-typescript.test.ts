import {SynthUtils,} from "@aws-cdk/assert";
import * as cdk from "@aws-cdk/core";
import {CdkTypescriptStack} from "../src/cdk-typescript-stack";

test("Empty Stack", () => {
  //const app = new cdk.App();
  //// WHEN
  //const stack = new CdkTypescript.CdkTypescriptStack(app, 'MyTestStack');
  //// THEN
  //expectCDK(stack).to(matchTemplate({
  //  "Resources": {}
  //}, MatchStyle.EXACT))
});

test("Stack matches snapshot", () => {
  const app = new cdk.App();
  //const stack = new CdkTypescript.CdkTypescriptStack(app, 'TestStack');
  const stack = new CdkTypescriptStack(app, 'MyTestStack');
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
