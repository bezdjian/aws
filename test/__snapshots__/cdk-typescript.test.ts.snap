// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack matches snapshot 1`] = `
Object {
  "Outputs": Object {
    "HelloFnApiEndpoint86B12587": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "HelloFnApi96F71B69",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "HelloFnApiDeploymentStageprod0787A28A",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters2101ff493c21785bd8f613f41401d505b01429a854f67ef9787d6e95a69d1dc1ArtifactHash66D920EE": Object {
      "Description": "Artifact hash for asset \\"2101ff493c21785bd8f613f41401d505b01429a854f67ef9787d6e95a69d1dc1\\"",
      "Type": "String",
    },
    "AssetParameters2101ff493c21785bd8f613f41401d505b01429a854f67ef9787d6e95a69d1dc1S3BucketF7D6FE0D": Object {
      "Description": "S3 bucket for asset \\"2101ff493c21785bd8f613f41401d505b01429a854f67ef9787d6e95a69d1dc1\\"",
      "Type": "String",
    },
    "AssetParameters2101ff493c21785bd8f613f41401d505b01429a854f67ef9787d6e95a69d1dc1S3VersionKey1833908F": Object {
      "Description": "S3 key for asset version \\"2101ff493c21785bd8f613f41401d505b01429a854f67ef9787d6e95a69d1dc1\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CdkDynamoDBTable7FCBB52E": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "TableName": "cdkTable",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "HelloFn7942DF23": Object {
      "DependsOn": Array [
        "HelloFnServiceRoleDefaultPolicy23D39CE4",
        "HelloFnServiceRoleDA97FE56",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters2101ff493c21785bd8f613f41401d505b01429a854f67ef9787d6e95a69d1dc1S3BucketF7D6FE0D",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters2101ff493c21785bd8f613f41401d505b01429a854f67ef9787d6e95a69d1dc1S3VersionKey1833908F",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters2101ff493c21785bd8f613f41401d505b01429a854f67ef9787d6e95a69d1dc1S3VersionKey1833908F",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "DB_TABLE": Object {
              "Ref": "CdkDynamoDBTable7FCBB52E",
            },
          },
        },
        "FunctionName": "CdkHelloFn",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "HelloFnServiceRoleDA97FE56",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "HelloFnApi96F71B69": Object {
      "Properties": Object {
        "Description": "Api gateway created by CDK for HelloFn Lambda",
        "Name": "HelloFnApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "HelloFnApiAccount2463B4FC": Object {
      "DependsOn": Array [
        "HelloFnApi96F71B69",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "HelloFnApiCloudWatchRoleD9669B2E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "HelloFnApiCloudWatchRoleD9669B2E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "HelloFnApiDeployment7DFA20C0e2701561287dd9dcbfe82c0b5c193389": Object {
      "DependsOn": Array [
        "HelloFnApiGET909E9DD6",
        "HelloFnApiscan030249A9",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "HelloFnApi96F71B69",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "HelloFnApiDeploymentStageprod0787A28A": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "HelloFnApiDeployment7DFA20C0e2701561287dd9dcbfe82c0b5c193389",
        },
        "RestApiId": Object {
          "Ref": "HelloFnApi96F71B69",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "HelloFnApiGET909E9DD6": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "HelloFn7942DF23",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "HelloFnApi96F71B69",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "HelloFnApi96F71B69",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "HelloFnApiGETApiPermissionMyTestStackHelloFnApiB884D7F8GET5DFCFEC8": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "HelloFn7942DF23",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "HelloFnApi96F71B69",
              },
              "/",
              Object {
                "Ref": "HelloFnApiDeploymentStageprod0787A28A",
              },
              "/GET/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "HelloFnApiGETApiPermissionTestMyTestStackHelloFnApiB884D7F8GET84F88627": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "HelloFn7942DF23",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "HelloFnApi96F71B69",
              },
              "/test-invoke-stage/GET/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "HelloFnApiscan030249A9": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "HelloFnApi96F71B69",
            "RootResourceId",
          ],
        },
        "PathPart": "scan",
        "RestApiId": Object {
          "Ref": "HelloFnApi96F71B69",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "HelloFnServiceRoleDA97FE56": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "HelloFnServiceRoleDefaultPolicy23D39CE4": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "CdkDynamoDBTable7FCBB52E",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "HelloFnServiceRoleDefaultPolicy23D39CE4",
        "Roles": Array [
          Object {
            "Ref": "HelloFnServiceRoleDA97FE56",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
